### ISSUE TITLE
Symlink (soft link) creation during `clone` fails on Windows with `[WinError 1314] A required privilege is not held by the client`

### ISSUE THREAD

**Issue by pwintz** [langfield/ki#37]
When I run `ki clone`, I see several errors that say `[WinError 1314] A required privilege is not held by the client`. Tracking down which notes caused the errors, I see that each note has cards in two (or more) different decks. This is not something I've ever done intentionally, but obviously, it can happen, so we need to handle it when it does come up.

The way Ki is designed, it seems to assume that all of a card's notes are in the same deck. I think this is a reasonable requirement, but we need to let users know, both in the documentation and with a warning when a card is skipped during cloning.

**Comment by langfield** [comment]
Great catch! So I've thought a lot about this problem. Most functions in ki have been modified to handle notes whose cards are split across multiple decks, and the way we write files to the filesystem reflects this.

You can see the symlinks used to handle this issue in a fresh clone of the example deck you sent me, or at least it works properly on POSIX. :sweat_smile: 

There are a couple of comments/TODOs littered throughout the codebase that still reflect the old state, and I believe the `ColNote` object still has a `deck` field, but the actual card files no longer have this field in the markdown source.

**Comment by langfield** [comment]
And you can see we even have a (simple) test for supporting this:

https://github.com/langfield/ki/blob/d75ef84867d11afd8ba2e4d84811e1a892a31ee7/tests/test_integration.py#L642-L649

**Comment by langfield** [comment]
> When I run `ki clone`, I see several errors that say `[WinError 1314] A required privilege is not held by the client`.

Can I see a full stack trace for one of these? Or a link to another comment if something like this has already been posted?

**Comment by langfield** [comment]
There are some possible related issues/causes mentioned in the `DEVELOPMENT` notes:

https://github.com/langfield/ki/blob/d75ef84867d11afd8ba2e4d84811e1a892a31ee7/DEVELOPMENT#L108-L115

https://github.com/langfield/ki/blob/d75ef84867d11afd8ba2e4d84811e1a892a31ee7/DEVELOPMENT#L290-L313

**Comment by langfield** [comment]
It appears to be related to symlink creation on Windows. See https://github.com/AcademySoftwareFoundation/rez/issues/594#issue-433869332.

> Presumably, because users on Windows can't (soft) symlink files without administrative privileges.
> 
> * One solution might be to simply not symlink Python, but just copy it.
> * Another is to use `hardlink` rather than `softlink` which does work without admin privs, but doesn't work across mounts, e.g. from `C:\source.txt` to `D:\dest.txt`.

**Comment by pwintz** [comment]
Here's the output when I run `test_integration.py::test_clone_handles_cards_from_a_single_note_in_distinct_decks
` (this also gives you the full stacktrace you requested).

```
C:\Users\pwintz\flashcards\ki\venv\Scripts\python.exe "C:\Program Files\JetBrains\PyCharm Community Edition 2021.1.2\plugins\python-ce\helpers\pycharm\_jb_pytest_runner.py" --target test_integration.py::test_clone_handles_cards_from_a_single_note_in_distinct_decks
Testing started at 2:37 AM ...
C:\Program Files\JetBrains\PyCharm Community Edition 2021.1.2\plugins\python-ce\helpers\pycharm\_jb_pytest_runner.py:28: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
Launching pytest with arguments test_integration.py::test_clone_handles_cards_from_a_single_note_in_distinct_decks --no-header --no-summary -q in C:\Users\pwintz\flashcards\ki\tests

  if version.LooseVersion(pytest.__version__) >= version.LooseVersion("6.0"):
============================= test session starts =============================
collecting ... collected 1 item

2022-08-11 02:38:00.236 | WARNING  | ki:write_decks:1296 - Failed to write card with cid:1651266810254.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1294, in write_decks
    F.symlink(note_path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\top\\a\\a.md' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\b\\a_Card 2.md'

2022-08-11 02:38:00.240 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\_media\\bullhorn-lg.png' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\_media\\bullhorn-lg.png'

2022-08-11 02:38:00.241 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\_media\\1sec.mp3' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\_media\\1sec.mp3'

2022-08-11 02:38:00.242 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\_media\\_vonNeumann.jpg' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\_media\\_vonNeumann.jpg'

2022-08-11 02:38:00.243 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\..\\top\\_media\\bullhorn-lg.png' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\a\\_media\\bullhorn-lg.png'

2022-08-11 02:38:00.243 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\..\\top\\_media\\1sec.mp3' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\a\\_media\\1sec.mp3'

2022-08-11 02:38:00.244 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\..\\top\\_media\\_vonNeumann.jpg' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\a\\_media\\_vonNeumann.jpg'

2022-08-11 02:38:00.245 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\..\\top\\_media\\1sec.mp3' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\b\\_media\\1sec.mp3'

2022-08-11 02:38:00.246 | WARNING  | ki:write_decks:1345 - Failed to create symlink to media.
Traceback (most recent call last):
  File "C:\Users\pwintz\flashcards\ki\ki\__init__.py", line 1343, in write_decks
    F.symlink(path, target)
  File "<@beartype(ki.functional.symlink) at 0x22166c84d30>", line 50, in symlink
  File "C:\Users\pwintz\flashcards\ki\ki\functional.py", line 147, in symlink
    os.symlink(target, path)
OSError: [WinError 1314] A required privilege is not held by the client: '..\\..\\..\\top\\_media\\_vonNeumann.jpg' -> 'C:\\Users\\pwintz\\AppData\\Local\\Temp\\pytest-of-pwintz\\pytest-27\\test_clone_handles_cards_from_0\\tmp_sh1k0oz\\split\\top\\b\\_media\\_vonNeumann.jpg'

2022-08-11 02:38:00.562 | DEBUG    | tests.test_integration:test_clone_handles_cards_from_a_single_note_in_distinct_decks:650 - C:\Users\pwintz\AppData\Local\Temp\pytest-of-pwintz\pytest-27\test_clone_handles_cards_from_0\tmp_sh1k0oz\split

tests\test_integration.py:644 (test_clone_handles_cards_from_a_single_note_in_distinct_decks)
tmp_path = WindowsPath('C:/Users/pwintz/AppData/Local/Temp/pytest-of-pwintz/pytest-27/test_clone_handles_cards_from_0')

    def test_clone_handles_cards_from_a_single_note_in_distinct_decks(tmp_path):
        col_file = get_split_col_file()
        runner = CliRunner()
        with runner.isolated_filesystem(temp_dir=tmp_path):
            clone(runner, col_file)
            logger.debug(os.path.abspath(SPLIT_REPODIR))
>           assert os.path.islink(Path(SPLIT_REPODIR) / "top" / "b" / "a_Card 2.md")
E           AssertionError: assert False
E            +  where False = <function islink at 0x0000022160B96040>((((WindowsPath('split') / 'top') / 'b') / 'a_Card 2.md'))
E            +    where <function islink at 0x0000022160B96040> = <module 'ntpath' from 'C:\\Program Files\\Python39\\lib\\ntpath.py'>.islink
E            +      where <module 'ntpath' from 'C:\\Program Files\\Python39\\lib\\ntpath.py'> = os.path
E            +    and   WindowsPath('split') = Path('split')

test_integration.py:651: AssertionError






============================== 1 failed in 1.10s ==============================

Process finished with exit code 1
```

**Comment by langfield** [comment]
[This](https://stackoverflow.com/a/59761201) also appears to be relevant.

# 2020+ TL;DR Answer
1. Enable "Developer Mode" in Windows 10/11 -- gives mklink permissions
2. Ensure symlinks are enabled in git with (at least) one of
    - **System setting:** check the checkbox when installing msysgit
    - **Global setting:** `git config --global core.symlinks true`
    - **Local setting:** `git config core.symlinks true`

---
Be careful, support for symlinks in git on Windows is relatively new. There are some bugs that still affect some git clients. Notably, symlinks with relative (..) paths are mangled in some programs because of a [(fixed)](https://github.com/libgit2/libgit2/pull/5435) [regression in libgit2](https://github.com/libgit2/libgit2/issues/5424). For instance, [GitKraken](https://www.gitkraken.com/) is affected by this because they are waiting on [nodegit](https://github.com/nodegit/nodegit) to update [libgit2](https://github.com/libgit2/libgit2) from v0.x (regression) to v1.x (fixed).
---
## Recreate missing/broken symlinks
Various levels of success have been reported across multiple git clients with one of these (increasingly forceful and "dangerous") options

Checkout: `git checkout -- path/to/symlink`
Restore (since git v2.23.0): `git restore -- path/to/symlink`
Switch branches (away and back)
Hard Reset: `git reset --hard`
Delete local repository and clone again
## Troubleshooting
`git config --show-scope --show-origin core.symlinks` will show you the level (aka "scope") the setting is set, where the configuration file (aka "origin") that is persisting it is, and the current value of the setting. Most likely a "local" configuration is overriding the "global" or "system" setting. `git config --unset core.symlinks` will clear a "local" setting allowing a higher level setting to take effect.

**Comment by pwintz** [comment]
It looks like my git configuration has `core.symlinks=false`:
```
$ git config core.symlinks
false
```

**Comment by pwintz** [comment]
Changing it to true looks like it had no effect.

**Comment by langfield** [comment]
@pwintz Yeah so this really worries me, though. I *really* don't want users to have to mess with their OS-level settings just to use this tool. If it eventually gets bundled into an add-on, and there are nascent plans to do exactly that, then this is completely untenable.

I don't see a light at the end of the tunnel yet. Perhaps some complicated mess that converts between hardlinks and softlinks using pre-commit hooks or something like that. :disappointed:

**Comment by langfield** [comment]
**An open question:** is it possible to handle (1) cards from a single note split across multiple decks, and (2) media deduplication, *without* using symlinks? And in particular, in a way that works on *Windows* :window:.

**Comment by pwintz** [comment]
What if the `*Card 2.md` file just is a plain text file with a format recognized by Ki that points to the "Card 1" file? Then, all of the note information could be placed in "Card 1", and Ki would use "Card 2" to correctly place cards but would use the "Card 1" file for the actual contents of the note. 

I think this would basically amount to doing a poor-man's implementation of symlinks within Ki, which might be a bit messy, but it would avoid the difficulties of making proper symlinks work on Windows.

**Comment by langfield** [comment]
**Note.** This issue is mitigated but not fixed by #38.

**Comment by langfield** [comment]
> What if the `*Card 2.md` file just is a plain text file with a format recognized by Ki that points to the "Card 1" file? Then, all of the note information could be placed in "Card 1", and Ki would use "Card 2" to correctly place cards but would use the "Card 1" file for the actual contents of the note.
> 
> I think this would basically amount to doing a poor-man's implementation of symlinks within Ki, which might be a bit messy, but it would avoid the difficulties of making proper symlinks work on Windows.

I like this a lot, and I think it is the road to an optimal solution! I *do* want ordinary symlinks to work on POSIX machines. Being able to open the card file and edit the single source of ground-truth is just too elegant of a solution to give up. I believe `F.symlink()` will have to be rewritten to check `sys.platform`, and we'll have to do some fancy footwork to convert back and forth between symlinks on POSIX and the files you described above on Windows. Perhaps there are existing solutions out there. I'll look into it.

**Comment by langfield** [comment]
One way this could be done is using Windows shortcuts. All `clone` calls on Windows will create shortcuts instead of soft links. We could use `pywin32`, which is very mature. See [here](https://stackoverflow.com/a/60944178).

> first, install the requirements
> 
> ```
> pip install pywin32
> pip install winshell
> ```
> then this is the code you have to write.
> 
> ```
> import os, winshell
> from win32com.client import Dispatch
> 
> path = r"H:\happy\hi\new.lnk"  # Path to be saved (shortcut)
> target = r"D:\New folder\new.exe"  # The shortcut target file or folder
> work_dir = r"D:\New folder"  # The parent folder of your file
> 
> shell = Dispatch('WScript.Shell')
> shortcut = shell.CreateShortCut(path)
> shortcut.Targetpath = target
> shortcut.WorkingDirectory = work_dir
> shortcut.save()
> ```
> for more details: https://winshell.readthedocs.io/en/latest/shortcuts.html

Note that `winshell` is unmaintained and old.

**Comment by langfield** [comment]
Alternatively, we could simply try to generate POSIX-style symlinks and use *those* in the same way we would use the files you described.

My real worry is how Windows users and POSIX users will manage collaborating on the same remote repository. This could be tough, and using POSIX-style symlinks (that don't actually "work" on Windows) would solve this problem. But I don't know if Windows preserves/supports the necessary file attributes.

**Comment by langfield** [comment]
So this is the WIP solution:

https://github.com/langfield/ki/blob/16887686c7b220ac4c938e4a3f7a665fb4c0de80/ki/__init__.py#L1894-L1907

**Comment by langfield** [comment]
@pwintz Hey, just checking in to check what your priorities are for Windows support. We got #38 merged into `main`, but this issue is still not resolved. Is this still your main concern, or is there something else more pressing to get working? Just want to ensure I'm prioritizing correctly! :smile:

**Comment by pwintz** [comment]
Hey @langfield, this isn't urgent for me now that #38 is merged. I never intentionally put cards from a single note into multiple decks. 
And sorry for being less engaged the last few days. I've been spending more time on my **actual** work haha. I'll circle back around and test out Ki more eventually!

**Comment by langfield** [comment]
@pwintz Sounds good! Yes, actual work is more important!

**Comment by langfield** [comment]
> Unfortunately, it poses an annoying problem for us, because whenever the user runs `git add` on these files, they will undo our hard work.

I've just realized that `git` for Windows must handle this issue somehow, because it isn't specific to our use-case. This should be tried-out on a Windows box to see what subsequent `git add` ops do.

**Comment by langfield** [comment]
@pwintz This has been fixed. Let me know if the issue persists!
